src/client/conn.rs

```diff
@@ -178,6 +148,127 @@ impl Connection {
+    ) -> error::Result<Transport<TcpStream>> {
+        let stream = Self::new_stream(config).await?;
+        let framed = IrcCodec::new(config.encoding())?.framed(stream);
+
+        Ok(Transport::new(&config, framed, tx))
+    }
+
+    #[cfg(feature = "tls-native")]
+    async fn new_secured_transport(
+        config: &Config,
+        tx: UnboundedSender<Message>,
+    ) -> error::Result<Transport<TlsStream<TcpStream>>> {
+        let mut builder = TlsConnector::builder();
+
+        if let Some(cert_path) = config.cert_path() {
+            let mut file = File::open(cert_path)?;
+            let mut cert_data = vec![];
+            file.read_to_end(&mut cert_data)?;
+            let cert = Certificate::from_der(&cert_data)?;
+            builder.add_root_certificate(cert);
+            log::info!("Added {} to trusted certificates.", cert_path);
+        }
+
+        if let Some(client_cert_path) = config.client_cert_path() {
+            let client_cert_pass = config.client_cert_pass();
+            let mut file = File::open(client_cert_path)?;
+            let mut client_cert_data = vec![];
+            file.read_to_end(&mut client_cert_data)?;
+            let pkcs12_archive = Identity::from_pkcs12(&client_cert_data, &client_cert_pass)?;
+            builder.identity(pkcs12_archive);
+            log::info!(
+                "Using {} for client certificate authentication.",
+                client_cert_path
+            );
+        }
+
+        let connector: tokio_tls::TlsConnector = builder.build()?.into();
+
+        let stream = Self::new_stream(config).await?;
+        let stream = connector.connect(config.server()?, stream).await?;
+        let framed = IrcCodec::new(config.encoding())?.framed(stream);
+
+        Ok(Transport::new(&config, framed, tx))
+    }
+
+    #[cfg(feature = "tls-rust")]
+    async fn new_secured_transport(
+        config: &Config,
+        tx: UnboundedSender<Message>,
+    ) -> error::Result<Transport<TlsStream<TcpStream>>> {
+        let mut builder = ClientConfig::default();
+        builder
+            .root_store
+            .add_server_trust_anchors(&TLS_SERVER_ROOTS);
+
+        if let Some(cert_path) = config.cert_path() {
+            let mut pem = BufReader::new(File::open(cert_path)?);
+            builder.root_store.add_pem_file(&mut pem).map_err(|_| {
+                error::Error::Io(Error::new(ErrorKind::InvalidInput, "invalid cert"))
+            })?;
+            log::info!("Added {} to trusted certificates.", cert_path);
+        }
+
+        if let Some(client_cert_path) = config.client_cert_path() {
+            let client_cert_pass = config.client_cert_pass();
+            let certs =
+                certs(&mut BufReader::new(File::open(client_cert_path)?)).map_err(|_| {
+                    error::Error::Io(Error::new(ErrorKind::InvalidInput, "invalid cert"))
+                })?;
+            let mut keys = rsa_private_keys(&mut BufReader::new(File::open(client_cert_pass)?))
```

Comment:
Is `client_cert_pass` a typo for `client_cert_path`? As far as I see, this patch-set doesn't seem to redefine `client_cert_pass` as a file-path. (I know, I'm reviewing this pretty late.)

Reply:
Oh good point...
The feature `tls-native` require a password stored in `client_cert_pass` while `tls-rust` require a path stored in `client_cert_pass` and will read the password stored on the file....
I personally think that it is cleaner to store the password in a file but for avoiding any breaking change we should follow the behavior of `tls-native`.